content/en/blog/_posts/2021-04-14-local-storage-features-go-beta.md

```diff
@@ -0,0 +1,205 @@
+permanently if storage capacity runs out on a node. The scheduler
+extenders try to avoid that, but cannot be 100% reliable.
+
+## Improvements in Kubernetes 1.21
+
+Starting with [external-provisioner
+v2.1.0](https://github.com/kubernetes-csi/external-provisioner/releases/tag/v2.1.0),
+released for Kubernetes 1.20, provisioning can be handled by
+external-provisioner instances that get [deployed together with the
+CSI driver on each
+node](https://github.com/kubernetes-csi/external-provisioner#deployment-on-each-node)
+and then cooperate to provision volumes ("distributed
+provisioning"). There is no need any more to have a central component
+and thus no need for communication between nodes, at least not for
+provisioning.
+
+A scheduler extender still needs some way to find out about capacity
+on each node. When PMEM-CSI switched to distributed provisioning in
+v0.9.0, this was done by querying the metrics data exposed by the
+local driver containers. But it is better also for users to eliminate
+the need for a scheduler extender completely. [Storage capacity
+tracking](/docs/concepts/storage/storage-capacity/),
+[introduced in
+1.19](https://kubernetes.io/blog/2020/09/01/ephemeral-volumes-with-storage-capacity-tracking/)
+and promoted to beta in Kubernetes 1.21, achieves that. It works by
+publishing information about capacity in `CSIStorageCapacity`
+objects. The scheduler itself then uses that information to filter out
+unsuitable nodes. Because information might be not quite up-to-date,
+pods may still get assigned to nodes with insufficient storage, it's
+just less likely and the next scheduling attempt for a pod should work
+better once the information got refreshed.
+
+So CSI drivers still need the ability to recover from a bad scheduling
+decision. That was one the reasons for the introduction of ["*generic*
+ephemeral
+volumes"](https://kubernetes.io/docs/concepts/storage/ephemeral-volumes/#generic-ephemeral-volumes),
+another related feature that got promoted to beta in 1.21. They are not
+limited to local storage and, as the name implies, also work for
+non-CSI storage drivers. The feature is based on automatically
+creating PVCs from a template in the pod spec and then lettings
+storage drivers react to those as if the PVC had been created
+manually.
+
+## Known limitations
+
+Both generic ephemeral volumes and storage capacity tracking increase
+the load on the API server. Whether that is a problem depends a lot on
+the kind of workload, in particular how many pods have volumes and how
+often those need to be created and destroyed.
+
+No attempt was made to model how scheduling decisions affect storage
+capacity. That's because the effect can vary considerably depending on
+how the storage system handles storage. The effect is that multiple
+pods with unbound volumes might get assigned to the same node even
+though capacity there is only sufficient for one pod. Scheduling
```

Comment:
"even though there is only sufficient capacity"

Reply:
Changed.